clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -4619,6 +4619,12 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
     // Buffer is a void**.
     Address Buf = EmitPointerWithAlignment(E->getArg(0));
 
+    if (getTarget().getTriple().getArch() == llvm::Triple::systemz) {
+      // Call LLVM's EH setjmp, which is lightweight.
```

Comment:
I'm a little concerned about modifying the semantics of the intrinsic like this.  I mean, the semantics are inherently a bit platform-specific, but this contradicts the documentation in llvm/docs/ExceptionHandling.rst.

If we need different semantics here for some reason, I'd prefer to use an intrinsic with a different name, to avoid any confusion.

Reply:
Hmm.  That's pretty messy.  See also 02269a66b362f1345cc12176f7e2cbc5474d2f02 and faa3abbe39a2b17546387fde04d2822e0402b418 .  I think what happened is that the meaning of the intrinsic was adapted to suit sjlj lowering, and nobody really thought about trying to expose a stable interface to frontends.

I still think it's a bit of a hazard to adjust the semantics of the intrinsic at this point without changing the name.  But there isn't any existing SystemZ bitcode to break, so it might be enough to just update the documentation.

Reply:
I see.  It seems back in the day the `eh.sjlj.setjmp` intrinsic was used by both the front-end to implement `__builtin_setjmp` and by the middle-end (SjLjEHPrepare.cpp) for exception handling.  Since these two users apparently require somewhat different information in the buffer, filling this in was left to the user.

However, these days those two users actually emit two different intrinsics.  The middle-end now uses `eh.sjlj.setup.dispatch` (which is apparently only intended for internal usage; it's not documented at all), while only the front-end continues to use `eh.sjlj.setjmp`.  See http://reviews.llvm.org/D9313.

I still think the most straightfoward way would be to have `eh.sjlj.setjmp` just be completely implemented in the back end, and simply do exactly what is needed to implement `__builtin_setjmp` on the platform, just like is done with longjmp.  This could be done for all targets (probably the cleanest solution, but would require updating the existing targets and might introduce incompatibilities with existing bytecode on those targets), or just for selected targets (I guess this would then need to be documented?  And should it be a positive or negative list of targets?).

Reply:
That makes sense.  I don't think there's any benefit to having the frontend to generate explicit stores.  I just want to make sure whatever solution we choose, it doesn't silently break code generated by other frontends (or older versions of clang).